algorithms/src/snark/marlin/data_structures/proof.rs

```diff
@@ -131,99 +134,136 @@ impl<F: PrimeField> Evaluations<F> {
+        size += serialized_vec_size_without_len(&self.g_b_evals, compress);
+        size += serialized_vec_size_without_len(&self.g_c_evals, compress);
         size
     }
 
     fn deserialize_with_mode<R: snarkvm_utilities::Read>(
-        batch_size: usize,
+        batch_sizes: &[usize],
         mut reader: R,
         compress: Compress,
         validate: Validate,
     ) -> Result<Self, snarkvm_utilities::SerializationError> {
-        let mut z_b_evals = Vec::with_capacity(batch_size);
-        for _ in 0..batch_size {
-            z_b_evals.push(CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?);
-        }
         Ok(Evaluations {
-            z_b_evals,
+            z_b_evals: batch_sizes
+                .iter()
+                .map(|batch_size| deserialize_vec_without_len(&mut reader, compress, validate, *batch_size))
+                .collect::<Result<Vec<Vec<_>>, _>>()?,
             g_1_eval: CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?,
-            g_a_eval: CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?,
-            g_b_eval: CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?,
-            g_c_eval: CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?,
+            g_a_evals: deserialize_vec_without_len(&mut reader, compress, validate, batch_sizes.len())?,
+            g_b_evals: deserialize_vec_without_len(&mut reader, compress, validate, batch_sizes.len())?,
+            g_c_evals: deserialize_vec_without_len(&mut reader, compress, validate, batch_sizes.len())?,
         })
     }
 }
 
 impl<F: PrimeField> Evaluations<F> {
-    pub(crate) fn from_map(map: &std::collections::BTreeMap<String, F>, batch_size: usize) -> Self {
-        let z_b_evals = map.iter().filter_map(|(k, v)| k.starts_with("z_b_").then_some(*v)).collect::<Vec<_>>();
-        assert_eq!(z_b_evals.len(), batch_size);
-        Self { z_b_evals, g_1_eval: map["g_1"], g_a_eval: map["g_a"], g_b_eval: map["g_b"], g_c_eval: map["g_c"] }
+    pub(crate) fn from_map(
+        map: &std::collections::BTreeMap<String, F>,
+        batch_sizes: BTreeMap<CircuitId, usize>,
+    ) -> Self {
+        let mut z_b_evals_collect: BTreeMap<CircuitId, Vec<F>> = BTreeMap::new();
+        let mut g_a_evals = Vec::with_capacity(batch_sizes.len());
+        let mut g_b_evals = Vec::with_capacity(batch_sizes.len());
+        let mut g_c_evals = Vec::with_capacity(batch_sizes.len());
+
+        for (label, value) in map {
+            if label == "g_1" {
+                break;
+            }
+
+            let circuit_id = CircuitId::from_witness_label(label);
+            if label.contains("z_b_") {
+                if let Some(z_b_i) = z_b_evals_collect.get_mut(&circuit_id) {
+                    z_b_i.push(*value);
+                } else {
+                    let mut values = Vec::with_capacity(batch_sizes[&circuit_id]);
```

Comment:
do these collections later get extended? at this point only one element is inserted into them

Reply:
Yes the loop this is running in should fill them all